mysys/buffered_error_log.h

```diff
@@ -0,0 +1,95 @@
+
+#pragma once
+
+#include <stdio.h>
+#include <cstring>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+extern char *buffered_error_log_filename;
+
+/** Stores log messages in a fixed size buffer, which can be dumped by large
+ * chunks instead of line by line, to increase performance for high throughput
+ * scenarios.
+ *
+ * The buffer is also dumped by the crash reporting code and during shutdown to
+ * ensure that nothing is lost, only delayed.
+ */
+class buffered_error_logger {
+  // We are using a pointer to string to guarantee that we can decrease the size
+  // of the buffer
+  using data_t = std::unique_ptr<std::string>;
+  // memory buffer storing the log messages
+  std::mutex data_mtx;
+  data_t data;
+
+ public:
+  void resize(std::size_t buffer_size) {
+    std::lock_guard<std::mutex> lk{data_mtx};
+
+    if (data.get() != nullptr && buffer_size == data->capacity()) {
+      return;
+    }
+
+    // Write out what we have currently, that way we don't have
+    // to deal with old data in the buffer
+    write_to_disk_();
+
+    if (buffer_size == 0) {
+      data.reset();
+      return;
+    }
+
+    data_t new_buffer(new std::string());
+    new_buffer->reserve(buffer_size);
+    data.swap(new_buffer);
+  }
+
+  ~buffered_error_logger() { write_to_disk(); }
+
+  void log(const char *msg, size_t len) {
+    std::lock_guard<std::mutex> lk{data_mtx};
+    if (data.get() == nullptr || data->capacity() == 0) return;
+    const auto msg_end = data->size() + len;
+    if (msg_end > data->capacity() - 1) {
```

Comment:
should we move implementation to .cc file?

Reply:
Done.